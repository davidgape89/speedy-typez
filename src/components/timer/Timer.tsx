import React, { useState, useEffect, forwardRef, useImperativeHandle } from 'react';

export enum TimerDirection {
  ASC = 1,
  DESC = -1,
}

interface TimerProps {
  initTime: number;
  onTimeOver: () => void;
  isRunning?: boolean;
  direction?: TimerDirection;
}

export interface TimerRef {
  reset: () => void;
}

export const Timer = forwardRef(({
    isRunning = true,
    initTime,
    direction = TimerDirection.DESC,
    onTimeOver
  }: TimerProps, ref) => {
  const [time, setTime] = useState(initTime);
  const [timeoutRef, setTimeoutRef] = useState();
  const minutes = Math.floor(time / 60).toString();
  const seconds = (time % 60).toString();

  if ( isRunning && !timeoutRef) {
    if (time === 0) {
      onTimeOver();
    } else {
      setTimeoutRef(
        setTimeout(() => {
          setTime(time => time + direction)
          setTimeoutRef(null);
        }, 1000)
      );
    }
  }

  useImperativeHandle(ref, () => ({
    reset() {
      setTime(initTime);
    }
  }));

  useEffect(() => {
    if (!isRunning && timeoutRef) {
      clearTimeout(timeoutRef);
      setTimeoutRef(null);
    }

    return () => clearTimeout(timeoutRef);
  }, [isRunning, timeoutRef]);

  return (
    <span>
      {minutes.length === 1 ? `0${minutes}`: minutes}:
      {seconds.length === 1 ? `0${seconds}`: seconds}
    </span>
  )
});
