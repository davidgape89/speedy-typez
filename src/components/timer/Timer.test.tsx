import React from 'react';
import {shallow} from 'enzyme';
import {Timer} from './Timer';

test('renders correctly', () => {
  const wrapper = shallow(<Timer initTime={60} onTimeOver={()=>{}}/>);

  expect(wrapper).toMatchSnapshot();
});
