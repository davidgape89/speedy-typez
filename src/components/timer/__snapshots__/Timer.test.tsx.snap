// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<span>
  01
  :
  00
</span>
`;
